Model/Gateway.php

```diff
@@ -0,0 +1,266 @@
+        }
+        if (is_null($reference) || strlen($reference) === 0) {
+            throw new \InvalidArgumentException("Reference is required");
+        }
+
+        if (function_exists('bcmul')) {
+            $int_amount = intval(bcmul($amount, 100));
+        } else {
+            $multiplied = round($amount * 100);
+            $int_amount = (int)$multiplied;
+        }
+
+        $payload = [
+            "transaction_id" => $transaction_id,
+            "amount" => $int_amount,
+            "reference" => $reference
+            ];
+
+        return $this->do_request("POST", "/refunds", $payload);
+    }
+
+    // TODO: auth/captures
+
+    /************** Private functions ***************/
+
+    /**
+     * Performs the request against the Fat Zebra gateway
+     * @param string $method the request method ("POST" or "GET")
+     * @param string $uri the request URI (e.g. /purchases, /credit_cards etc)
+     * @param array $payload the request payload (if a POST request)
+     * @return array
+     * @throws \Exception
+     */
+    private function do_request($method, $uri, $payload = null)
+    {
+        $curl = curl_init();
+        if (is_null($this->api_version)) {
+            $url = $this->url . $uri;
+            curl_setopt($curl, CURLOPT_URL, $url);
+        } else {
+            $url = $this->url . "/v" . $this->api_version . $uri;
+            curl_setopt($curl, CURLOPT_URL, $url);
+        }
+
+        $payload["test"] = $this->test_mode;
+
+        curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, true);
+        curl_setopt($curl, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+        curl_setopt($curl, CURLOPT_USERPWD, $this->username . ":" . $this->token);
+        curl_setopt($curl, CURLOPT_HTTPHEADER, ["User-agent: FatZebra Magento2 Library " . $this->version]);
+
+        if ($method == "POST" || $method == "PUT") {
+            curl_setopt($curl, CURLOPT_POST, true);
+            curl_setopt($curl, CURLOPT_HTTPHEADER, ["Content-type: application/json", "User-agent: FatZebra PHP Library " . $this->version]);
+            curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($payload));
+        }
+
+        if ($method == "PUT") {
+            curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "PUT");
+        }
+
+        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, true);
+        curl_setopt($curl, CURLOPT_SSLVERSION, 6); // CURLOPT_SSLVERSION_TLSv1_2
+        curl_setopt($curl, CURLOPT_TIMEOUT, $this->timeout);
+
+        $data = curl_exec($curl);
+
+        if (curl_errno($curl) !== 0) {
+            if (curl_errno($curl) == 28) {
+                throw new TimeoutException("cURL Timeout: " . curl_error($curl));
+            }
+            throw new \Exception("cURL error " . curl_errno($curl) . ": " . curl_error($curl));
+        }
+        curl_close($curl);
+
+        $response =  json_decode($data, true);
```

Comment:
Is it worth (now that we're in Magento context) moving towards \Magento\Framework\Serialize\Serializer\Json?